src/createReducer.ts

```diff
@@ -8,29 +8,40 @@ function composeMiddleware(chain) {
   };
 }
 
-const createReducer = (...middlewares) => (reducer, initialState, initializer = value => value) => {
-  const ref = useRef(initializer(initialState));
-  const [, setState] = useState(ref.current);
-  let middlewareDispatch = (_ = {}) => {
-    throw new Error(
-      'Dispatching while constructing your middleware is not allowed. ' +
-        'Other middleware would not be applied to this dispatch.'
+const createReducer = (...middlewares) => {
+  const composedMiddleware = composeMiddleware(middlewares);
+
+  return (reducer, initialState, initializer = value => value) => {
+    const ref = useRef(initializer(initialState));
+    const dispatchRef = useRef((_ = {}) => {
+      throw new Error(
+        'Dispatching while constructing your middleware is not allowed. ' +
+          'Other middleware would not be applied to this dispatch.'
+      );
+    });
+    const [, setState] = useState(ref.current);
+
+    const dispatch = useCallback(
+      action => {
+        ref.current = reducer(ref.current, action);
+        setState(ref.current);
+        return action;
+      },
+      [reducer]
     );
+
+    useMemo(() => {
```

Comment:
These changes look great 👌

Can you explain why you use `useMemo` instead `useEffect` here?

Reply:
Could we use `useIsomorphicLayoutEffect` for this?

Reply:
@wardoost I simplified createReducer - now we no need any useMemo or useEffect (all useReducer functionality still available)